eip/EIPS/eip-4337.md

```diff
@@ -58,45 +69,88 @@ The core interface of the entry point contract is as follows:
 
 ```c++
 function handleOps
-    (UserOperation[] calldata ops, address payable beneficiary)
-    public;
+    (UserOperation[] calldata ops, address payable beneficiary
+     AggregatorInfo[] calldata aggregators, uint[] memory execOrder
+    ) public
 
 function simulateValidation
-    (UserOperation calldata userOp)
-    external returns (uint256 preOpGas, uint256 prefund) {
+    (UserOperation calldata userOp, aggregatedSignature)
+    external returns (uint256 preOpGas, uint256 prefund)
+    
+struct AggregatorInfo {
+    IAggregator aggregator;
+    uint count;
+    bytes signature;
+}
 ```
 
 The core interface required for a wallet to have is:
 
-```c++
-function validateUserOp
-    (UserOperation calldata userOp, bytes32 requestId, uint256 missingWalletFunds)
-    external;
+```solidity
+interface IWallet {
+  function validateUserOp
+      (UserOperation calldata userOp, bytes32 requestId, uint256 missingWalletFunds)
+      external;
+}
+```
+The wallet
+* MUST validate the caller is a trusted EntryPoint
+* The requestId is a hash over the userOp (except signature), entryPoint and chainId
+* MUST validate the signature is a valid signature of the `requestId`
+* MUST pay the entryPoint (caller) at least the "missingWalletFunds" (which might be zero, in case current wallet's deposit is high enough)
+* the wallet MAY pay more than this minimum, to cover future transactions
+* Note that in case the request comes through an aggregator, the signature will be empty.
+
+A Wallet that supports aggregated signature should have the interface:
+```solidity
+interface IAggregatedWallet is IWallet {
+  function getAggregator() view returns (address);
+}
+```
+* **getAggregator()** returns the aggregator this wallet supports.
+* The wallet is guaranteed to that the EntryPoint will only call its `validateUserOp` (with no signature) after calling the aggregator's validateSignatures
+* The wallet should also support aggregator-specific getter (e.g. `getAggregationInfo()`).
+  This method should export the wallet's public-key to the aggregator, and possibly more info
+* validateUserOp shouldn't validate the signature (since it is verified by the aggregator),
+  but it MUST verify the signature field is shorter than 63 bytes
```

Comment:
Is there a way for a wallet to support both individual ERC-4337 transactions and aggregated-singature ERC-4337 transactions? Because with what you define here the wallet either accepts no signature, or must check signature, but not both.

It can call `IAggregator(this.getAggregator()).validateSignatures(address(msg.sender), [userOp], aggregatedSignature);` internally if there is a signature to avoid duplication.

Also, `shorter than 63 bytes` is super confusing - why not use zero length?

Reply:
fixed with the new API